### ISSUE TITLE
Compiler quits SILENTLY when compiling code with generic types.

### ISSUE THREAD

**Issue by tommo** [nim-lang/Nim#20416]
### What happened?

Simplified example:

```nim
import std/[tables]

type
  Item[T] = object
    link:ptr Item[T]
    data:T

  Cell[T] = object
    inner:ptr Item[T]

  Container*[T] = object
    a:Table[ int, ptr Cell[T]]

proc foo( sg:Container ) =   # <-- adding [T] parameter can make it compile.
  discard
```
The compiler( devel version ) quits silently when compiling the code above.

EDIT: calling fixed generic proc ( adding generic parameter ) could cause the compiler to quit without error hints.

```nim
import std/[tables]

type
  Item[T] = object
    link:ptr Item[T]
    data:T

  Cell[T] = object
    inner:ptr Item[T]

  Container*[T] = object
    a:Table[ int, ptr Cell[T]]

proc foo[T]( sg:Container[T] ) = # <- this compiles
  discard

var x:Container[int]
foo(x) # <- quits silently here
```

### Nim Version

Nim Compiler Version 1.7.1 [MacOSX: arm64]
Compiled at 2022-09-23
Copyright (c) 2006-2022 by Andreas Rumpf

git hash: 7739e23420c9a7a4ec7eccdddc0a39df9e905aa8
active boot switches: -d:release

### Current Standard Output Logs

exit code -11
no error information

**Comment by beef331** [comment]
[#be4bd8a0edd527b24679372b8cb9d2afa548056d](https://github.com/nim-lang/Nim/commit/be4bd8a0edd527b24679372b8cb9d2afa548056d) Seems to be the culprit.

```nim
import std/[tables]

type
  Item[T] = ref obzject
    link: Item[T]
    data:T

  Cell[T] = ref object
    inner: Item[T]

  Container*[T] = object
    a:Table[ int, Cell[T]]

proc foo( sg:Container ) =   # <-- adding [T] parameter can make it compile.
  discard

Container[int]().foo
```
Compiles so seems @deech's changes miss fields that are `ptr object` or `ref object`

**Comment by tommo** [comment]
Anyone, please have a look at this? This regression has been blocking some parts of my progress for a while. Thanks!

**Comment by Araq** [comment]
@deech this is a regression caused by your fix for #20348

**Comment by deech** [comment]
Sorry, I will look into this asap.
